RavenDB-24211 add parameters to the ai agent configuration

karmeli87 opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-24211

### Additional description

For better UX and studio work we require to have parameters in order to validate that we fill the agent config properly and that we start the chat with all the required parameters.

Bonus: we also have in this PR some preparation for a strongly typed API

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [x] New feature

### How risky is the change?

- [x] Low 
- [ ] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing
- [ ] Existing tests verify the correct behavior

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [ ] No

### UI work

- [x] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [ ] No UI work is needed

ayende commented on src/Raven.Client/Documents/Operations/AI/Agents/AiAgentConfiguration.cs:
```diff
@@ -80,6 +80,8 @@ public AiAgentConfiguration(string connectionStringName, string systemPrompt)
     /// </summary>
     public PersistenceConfiguration Persistence { get; set; }
 
+    public List<string> Parameters { get; set; } = [];
```
Why isn't that already a `HashSet` ?

ayende commented on src/Raven.Server/Documents/Handlers/AI/Agents/AiAgentProcessorForAddOrUpdateAiAgent.cs:
```diff
@@ -41,4 +49,36 @@ public override async ValueTask ExecuteAsync()
         }
     }
+
+    private static void ValidateConfiguration(JsonOperationContext context, AiAgentConfiguration configuration)
```
This looks really awkward way of validating this. I had to read through that a couple of times to understand what this is doing. 

You may have queries that aren't using all the parameters - for example - "product catalog search".

karmeli87 replied:
I indeed had fun here :)

The point here is that we care only about the parameters that the user has to provide us (the scope)
We shouldn't validate the parameters that are used by the LLM like the $query for the "product catalog search"

ayende replied:
I don't follow.

I _think_ that what you are trying to do here is to verify that in the query:
- Either the parameters are defined in the schema
- They are defined as global agent parameters

But the code is really hard to follow

ppekrol replied:
Wouldn't it be simpler to Union conf parameters + tool parameters, then all of the query parameters should be in that set?

ayende replied:
Yes, that is what I would generally expect.

ppekrol replied:
I spoke with @karmeli87 directly. He knows what to do.

karmeli87 committed:
RavenDB-24211 add parameters to the ai agent configuration

- validate proper agent configuration
- validate required parameters to start a chat

ppekrol commented on src/Raven.Server/Documents/Handlers/AI/Agents/AiAgentProcessorForAddOrUpdateAiAgent.cs:
```diff
@@ -41,4 +51,45 @@ public override async ValueTask ExecuteAsync()
+                }
+            }
+        }
+
```
check also llmParams does not contain scopeParams

karmeli87 committed:
RavenDB-24211 address PR comments

karmeli87 committed:
RavenDB-24211 address PR comments
